### PR TITLE
refactor: upgrade AutocompleteJS to v1 & update related code (Top Navigation Desktop Only)

### PR DESCRIPTION
**Pull request by thecristen**
**Asana Tickets:**
- [Upgrade AutocompleteJS to v1 & update related code (Top Navigation Desktop Only)](https://app.asana.com/0/385363666817452/1203649294497429/f)
- [🔎 Search | Header search is posting three times for each search instead of once](https://app.asana.com/0/385363666817452/1203042792105778/f)
- [🔎 Search | Don't search for one character unless it's a number](https://app.asana.com/0/0/1203466753992496/f)

# Backend highlights

> [!IMPORTANT]
> Our search box (both old and new) works by sending a POST request to our `/search/query` endpoint, which in turn does some processing and then encodes and sends a rather complex queries request to Algolia.

Before, the shaping the requests data for sending to Algolia was mostly being done with a mess of JS class functions, and finished in the backend in the `SearchController`. Now, this work is all shifted to the backend, to keep the frontend JS simpler. This is mostly achieved though the new `Algolia.Query.Request` module. 

----

> [!NOTE]
> This PR introduces our first Phoenix-based functional component. 

`<.algolia_autocomplete />` is intended to be reused wherever we'll need a search box. While it depends on LiveView under the hood to handle instantiation via a `phx-hook` object, it's a simpler component with no lifecycle or state. When mounted, it automatically calls `setupAlgoliaAutocomplete()`.

----

# Frontend highlights

> [!WARNING]
> There's a lot of JS here, but almost all of it is just here to correctly [specify the many parameters supported by the autocomplete-js library](https://www.algolia.com/doc/ui-libraries/autocomplete/api-reference/autocomplete-js/autocomplete/#parameters). 

Because of this, I tried to be careful with unit testing. We can assume the library itself is adequately tested, so the tests included here focus on portions that are unique to our particular implementation.

----

> [!NOTE]
> Three "sources" are searchable in the header search box

- geolocation (prompts user browser to share location), 
- locations in general (calls a backend endpoint that makes the request to AWS Location Service)
- our actual Algolia indexes (calls the backend endpoint noted in the "Backend highlights" section above)

None of this is newly implemented here, we're reusing endpoints and helper functions and templates as much as possible.

----

> [!WARNING]
> The autocomplete-js library typings are a bit chaotic.

Dealing with multiple sources (which each return results of different sorts of object) made typing more complicated still. Fighting with TypeScript was a big part of this work; I tried my best!

----

Visually, I aimed to keep everything as close to our current implementation as possible. I built off the library's default theme to help me achieve that, so there's some custom CSS based on that.

----

I tweaked the current header navigation and integration testing code to ensure compatibility.

---

#### General checks
* [x] **Are the changes organized into self-contained commits with descriptive and well-formatted commit messages?** This is a good practice that can facilitate easier reviews.
* [x] **Testing.** Do the changes include relevant passing updates to tests? This includes updating screenshots. Preferably tests are run locally to verify that there are no test failures created by these changes, before opening a PR.
* [x] **Tech debt.** Have you checked for tech debt you can address in the area you're working in? This can be a good time to address small issues, or create Asana tickets for larger issues.

#### New UI, or substantial UI changes
* [x] **Cross-browser compatibility** is less of an issue now that we're no longer supporting IE, but changes still need to work as expected in Safari, Chrome, and Firefox.
* [x] **Are interactive elements accessible?** This includes at minimum having [relevant keyboard interactions](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#keyboard-interaction) and [visible focus](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#visible-focus), but can also include verification with screen reader testing.
* [ ] **Other accessibility checks** such as sufficient [color constrast](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#color-contrast), or whether the layout holds up at 200% [zoom level](https://thinkdobecreate.com/articles/4-required-tests-before-shipping-new-features/#visible-focus).

### CODE REVIEW

File: apps/site/assets/js/algolia-result.js [hunk 1]
```diff
@@ -177,7 +177,7 @@ function _fileIcon(hit) {
   }
 }
 
-function _contentIcon(hit) {
+export function _contentIcon(hit) {
```
Comment (kotva006) [inline apps/site/assets/js/algolia-result.js:180]: nit-pick: For consistency this function should not start with an `_`

**Comment by thecristen** [comment]
> Will using a live view make writing front end tests harder? Is this something we need to mock when rendering component tests?

@kotva006 Probably not. Did you see the tests this PR has and do you think another type of test is needed? 

In this (relatively simple) case, the [frontend test](https://github.com/mbta/dotcom/pull/1781/files#diff-91c8edbb81bd0f59492edbe655100d1e24ab298c6896bd187a59930cb67e9d83) operates on HTML that's presumed already-generated from the backend, and assesses whether instantiating the frontend library on it works as expected. And the [backend test](https://github.com/mbta/dotcom/pull/1781/files#diff-8692b839200ed3496834006d4236e5962962f46e49f18a9b04dcdf603a41b821) checks for expected generated markup. And the [integration test](https://github.com/mbta/dotcom/pull/1781/files#diff-02c6e61c73744118872b666b4d0c2d5aac7ad37908a14e07d7fe634b518c3dd8) covers the is-it-present-if-you-go-visit-the-page piece.

For a more complex component that leverages more of LiveView's features around handling events and updating state and interacting with the server, I imagine those tests will be more (if not entirely) on the Elixir-side, using various tooling that comes with [LiveView](https://hexdocs.pm/phoenix_live_view/Phoenix.LiveViewTest.html). 🤷🏼‍♀️

File: apps/site/assets/ts/ui/autocomplete/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+export function isRouteItem(x: Item): x is RouteItem {
+  return Object.keys(x).includes("route");
+}
+
+export function isContentItem(x: Item): x is ContentItem {
+  return Object.keys(x).includes("_content_type");
+}
+
+export const getItemType = (item: Item): string => {
```
Comment (kotva006) [inline apps/site/assets/ts/ui/autocomplete/helpers.ts:17]: This function doesn't seem to be used anywhere

Comment (thecristen) [reply apps/site/assets/ts/ui/autocomplete/helpers.ts:17]: nice! will remove

**Review by kotva006 — APPROVED** [review]
Looks Good

**Pull request merged**